### DISCUSSION TITLE
Call from invalid thread when using ReactiveCommand.CreateFromTask

### DISCUSSION THREAD

**Discussion by CzBuCHi** [AvaloniaUI/Avalonia#11880] [Q&A]
**Describe the bug**
First couple of hours experimenting with Avalonia, so i may do things wrong ...

anyway: i have this view model and service:

```
public class LoginControlViewModel : ReactiveObject
{
    private readonly SessionService _SessionService;
    private string _Host = string.Empty;
    private string _Password = string.Empty;
    private string _UserName = string.Empty; 

    // ommited wrapper properties for string fields above

    public LoginControlViewModel(SessionService sessionService) {
        _SessionService = sessionService;
        Login           = ReactiveCommand.CreateFromTask(ExecuteLogin);
    }

    public ICommand Login { get; }

    private async Task ExecuteLogin() {
        if (!await _SessionService.Login(_Host, _UserName, _Password)) {
            // login failed
        }
    }
}

public class SessionService {
    private readonly HttpClient _HttpClient;

    public SessionService(HttpClient httpClient) {
        _HttpClient = httpClient;
    }

    public async Task<bool> Login(string host, string userName, string password) {
        var authorization = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{userName}:{password}"));

        var request = new HttpRequestMessage(HttpMethod.Get, host + "/jwt");
        request.Headers.Authorization = new("BASIC", authorization);

        HttpResponseMessage response;
        try {
            response = await _HttpClient.SendAsync(request).ConfigureAwait(false);
        } catch {
            return false;
        } 

        Session = response.IsSuccessStatusCode
            ? await response.Content.ReadAsStringAsync()
            : null;

        return Session != null;
    }
}
```
i get "Call from invalid thread" exception with stack trace bellow when i step over `await _HttpClient.SendAsync` part of my service which is not catched in my code and crash whole app...

note: exception is thrown before server responds but after http request is made (i have breakpoint in server code, that is hit with correct request)

**Stack**
```
   at Avalonia.Threading.Dispatcher.<VerifyAccess>g__ThrowVerifyAccess|16_0() in /_/src/Avalonia.Base/Threading/Dispatcher.cs:line 77
   at Avalonia.AvaloniaObject.GetValue[T](StyledProperty`1 property) in /_/src/Avalonia.Base/AvaloniaObject.cs:line 227
   at Avalonia.Controls.Button.get_Command()
   at Avalonia.Controls.Button.CanExecuteChanged(Object sender, EventArgs e)
   at ReactiveUI.ReactiveCommandBase`2.OnCanExecuteChanged(Boolean newValue)
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.IdentitySink`1.OnNext(T value)
   at System.Reactive.Subjects.FastImmediateObserver`1.EnsureActive(Int32 count)
   at System.Reactive.Subjects.FastImmediateObserver`1.EnsureActive()
   at System.Reactive.Subjects.ReplaySubject`1.ReplayBase.OnNext(T value)
   at System.Reactive.Subjects.ReplaySubject`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.IdentitySink`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.Linq.ObservableImpl.DistinctUntilChanged`2._.OnNext(TSource value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.Linq.ObservableImpl.CombineLatest`3._.SecondObserver.OnNext(TSecond value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.IdentitySink`1.OnNext(T value)
   at System.Reactive.Subjects.FastImmediateObserver`1.EnsureActive(Int32 count)
   at System.Reactive.Subjects.FastImmediateObserver`1.EnsureActive()
   at System.Reactive.Subjects.ReplaySubject`1.ReplayBase.OnNext(T value)
   at System.Reactive.Subjects.ReplaySubject`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.IdentitySink`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.Linq.ObservableImpl.DistinctUntilChanged`2._.OnNext(TSource value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.IdentitySink`1.OnNext(T value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.Linq.ObservableImpl.Select`2.Selector._.OnNext(TSource value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.Linq.ObservableImpl.Scan`2._.OnNext(TSource value)
   at System.Reactive.SafeObserver`1.WrappingSafeObserver.OnNext(TSource value)
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value)
   at System.Reactive.ObserveOnObserverLongRunning`1.Drain()
   at System.Reactive.ObserveOnObserverLongRunning`1.<>c.<.cctor>b__17_0(ObserveOnObserverLongRunning`1 self, ICancelable cancelable)
   at System.Reactive.Concurrency.DefaultScheduler.LongRunning.LongScheduledWorkItem`1.<>c.<.ctor>b__3_0(Object thisObject)
   at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl.<>c.<StartThread>b__8_0(Object itemObject)
```

**Expected behavior**
No exception

**Desktop**

- OS: windows 11 22H2
- Version 11.0.0-rc1.1

Any kick in right direction is appreciated ...

**Comment by maxkatz6** [comment]
Accessing UI elements from the non-UI thread is not supported.
In your case you can see that `Button.Command` property was used from a non-UI thread created at `ConcurrencyAbstractionLayerImpl.<>c.<StartThread>`.

I am not sure why ReactiveUI reads this property on HTTP request, but I assume it wants to update CanExecute value while async operation is in progress. Like, button should be disabled when it's in progress.

Can you double check if `.UseReactiveUI()` is used in your AppBuilder? It's needed to register Avalonia UI thread to be used from ReactiveUI.

**Reply by CzBuCHi** [reply]
TL/DR: yes i have called UseReactiveUI ....

long answer:
my program.cs looks like this:
```
using System;
using Avalonia;
using Avalonia.ReactiveUI;

internal static class Program
{
    // Initialization code. Don't use any Avalonia, third-party APIs or any
    // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
    // yet and stuff might break.
    [STAThread]
    public static void Main(string[] args) {
        BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);
    }

    // Avalonia configuration, don't remove; also used by visual designer.
    public static AppBuilder BuildAvaloniaApp() {
        return AppBuilder.Configure<App>()
                         .UsePlatformDetect()
                         .WithInterFont()!
                         .LogToTrace()
                         .UseReactiveUI();
    }
}
```
App.axaml.cs like this:
```
using System;
using System.Reflection;
using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Markup.Xaml;
using Celebrum.ViewModels;
using Celebrum.Views;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using ReactiveUI;
using Splat;
using Splat.Microsoft.Extensions.DependencyInjection;
using Splat.Microsoft.Extensions.Logging;

public class App : Application
{
    private readonly IHost _Host;

    public App() {
        _Host = Host.CreateDefaultBuilder()
                    .ConfigureServices((_, services) => {
                         services.UseMicrosoftDependencyResolver();
                         var resolver = Locator.CurrentMutable;
                         resolver.InitializeSplat();
                         resolver.InitializeReactiveUI();
                         Bootstrapper.ConfigureServices(services);
                     })
                    .ConfigureLogging(o => o.AddSplat())
                    .Build();
        _Host.Services.UseMicrosoftDependencyResolver();
    }

    public static IServiceProvider Container => ((App)Current!)._Host.Services;

    public override void Initialize() {
        AvaloniaXamlLoader.Load(this);
    }

    public override void OnFrameworkInitializationCompleted() {
        if (ApplicationLifetime! is IClassicDesktopStyleApplicationLifetime desktop) {
            desktop.MainWindow = new DialogWindow {
                ViewModel = new() {
                    Width         = 1000,
                    Height        = 550,
                    Title         = "Title",
                    CanResize     = true,
                    ShowInTaskbar = true,
                    Content       = Container.GetRequiredService<LoginControlViewModel>()
                }
            };

            desktop.Exit += (_, _) => _Host.Dispose();
        }

        base.OnFrameworkInitializationCompleted();
    }
}

```
Bootstrapper.cs is just simple DI registration calls:
```
using Microsoft.Extensions.DependencyInjection;

public static class Bootstrapper
{
    public static void ConfigureServices(IServiceCollection services) {
        _ = services.AddHttpClient();
        _ = services.AddSingleton<ISessionService, SessionService>();
        _ = services.AddTransient<LoginControlViewModel>();
    }
}
```
and csproj has just couple extra nugets:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <Nullable>enable</Nullable>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="Models\" />
    <AvaloniaResource Include="Assets\**" />
    <Folder Include="Services\" />
  </ItemGroup>


  <ItemGroup>
    <PackageReference Include="Avalonia" Version="11.0.0-rc1.1" />
    <PackageReference Include="Avalonia.Desktop" Version="11.0.0-rc1.1" />
    <PackageReference Include="Avalonia.Themes.Fluent" Version="11.0.0-rc1.1" />
    <PackageReference Include="Avalonia.Fonts.Inter" Version="11.0.0-rc1.1" />
    <!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
    <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.0.0-rc1.1" />
    <PackageReference Include="Avalonia.ReactiveUI" Version="11.0.0-rc1.1" />
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
    <PackageReference Include="Splat.Microsoft.Extensions.DependencyInjection" Version="14.6.37" />
    <PackageReference Include="Splat.Microsoft.Extensions.Logging" Version="14.6.37" />
  </ItemGroup>





</Project>

```

**Reply by maxkatz6** [reply]
This part looks suspicious:
```
                         var resolver = Locator.CurrentMutable;
                         resolver.InitializeSplat();
                         resolver.InitializeReactiveUI();
```

I wonder if it can override Avalonia ReactiveUI intergration

**Reply by CzBuCHi** [reply]
idk ... but when i removed it i got same result .... 
i will create minimal core to reproduce to remove my Pc from the equation ...

edit: 
of course when i created new project using communitytoolkit instead of reactiveui and without Microsoft.Extensions.DependencyInjection it worked as expected ...

so my guess is that real question is how to create avalonia + Microsoft.Extensions.DependencyInjection correctly ...

edit2:
in the end my DI bootstrap code looks like this
```
static App() {
        var services = new ServiceCollection();
        Bootstraper.RegisterServices(services);
        Provider = services.BuildServiceProvider();
    }

    public static ServiceProvider Provider { get; }

    public void Dispose() {
        Provider.Dispose();
    }
```
probably not good idea to dispose static property but i think App is singleton anyway ...

so i think problem is somewhere in reactiveui and not avalonia ....

**Reply by maxkatz6** [reply]
As an option, you can initialize host from OnFrameworkInitializationCompleted

**Comment by timunie** [comment]
I nowadays find that using the ComminityToolkit.MVVM RelayCommands are much easier to use. It works together with RxUI, so you can try to swap out that particular command which doesn't work well. Easiest is to use their source generators. 

https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/relaycommand#asynchronous-commands

**Reply by CzBuCHi** [reply]
im still in experimental phase where im trying to figure out what libraries to use - so starting again but with toolkit was not big deal ... (i picked reactiveui because it was preselected in template)

**Reply by timunie** [reply]
yeah it was the best back then when Avalonia started. Things have changed ^^. In v11, we will give you the choice via VS TemplateStudio.

**Reply by ColtonHoksbergen** [reply]
I am still having this issue even on the most basic uses of the community toolkit methodology.  I think I'm following the documentation correctly, but I cannot get around this issue.

**Reply by timunie** [reply]
@ColtonHoksbergen without seeing your code no option for help

**Comment by ColtonHoksbergen** [comment]
Sure, so here's my XAML:

<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:AvaloniaApplication2.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication2.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel"
        Icon="/Assets/avalonia-logo.ico"
        Title="AvaloniaApplication">

    <Design.DataContext>
        
        <vm:MainWindowViewModel/>
    </Design.DataContext>

	<Grid RowDefinitions="50, *">
		<StackPanel Orientation="Horizontal" HorizontalAlignment="Center">
			<Button Content="Add" Width="75" Height="30" Margin="0,0,5,0" Command="{Binding AddItem}"/>
			<Button Content="Delete" Width="75" Height="30" Margin="0,0,5,0" Command="{Binding DeleteItem}"/>
			<Button Content="Save" Width="75" Height="30" Margin="0,0,5,0" />
		</StackPanel>
		<Grid Grid.Row="1" ColumnDefinitions="7*, 3*">
			<DataGrid Margin="20" ItemsSource="{Binding Items}"
						SelectedItem="{Binding SelectedItem}"
						GridLinesVisibility="All"
						BorderThickness="1" BorderBrush="Gray">
				<DataGrid.Columns>
					<DataGridTextColumn Header="Name" Binding="{Binding Name}" Width="4*"/>
					<DataGridTextColumn Header="Serial Number" Binding="{Binding SerialNumber}" Width="4*"/>
					<DataGridTextColumn Header="Content" Binding="{Binding Quantity}" Width="2*"/>
				</DataGrid.Columns>
			</DataGrid>

			<StackPanel Grid.Column="1">
				<Label Content="Name" />
				<TextBox Margin="5,0,5,25" Text="{Binding SelectedItem.Name, UpdateSourceTrigger=PropertyChanged}" ></TextBox>
				<Label Content="Serial Number" />
				<TextBox Margin="5,0,5,25" Text="{Binding SelectedItem.SerialNumber, UpdateSourceTrigger=PropertyChanged}"></TextBox>
				<Label Content="Content" />
				<TextBox Margin="5,0,5,25" Text="{Binding SelectedItem.Quantity, UpdateSourceTrigger=PropertyChanged}"></TextBox>
			</StackPanel>
		</Grid>
	</Grid>

</Window>



And then in my MainWindowViewModel I have the following:
[RelayCommand]
public void AddItem()
{
    Items.Add(new Item { Name = "New Item", SerialNumber = "000", Quantity = 0 });
}


I have tried to wrap the action in an async call, as I was under the impression that might help, but still get the same result.

For further context, here's my program.cs:

using Avalonia;
using System;

namespace AvaloniaApplication2
{
    internal sealed class Program
    {
        // Initialization code. Don't use any Avalonia, third-party APIs or any
        // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
        // yet and stuff might break.
        [STAThread]
        public static void Main(string[] args) => BuildAvaloniaApp()
            .StartWithClassicDesktopLifetime(args);

        // Avalonia configuration, don't remove; also used by visual designer.
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .WithInterFont()
                .LogToTrace();
    }
}


I can provide any other information on request. My apologies, as I am new to the WPF/Avalonia world, so I may not understand what is all pertinent to the issue.

**Reply by timunie** [reply]
And Items is ObservableCollection?

You may also be interested in the TodoList we have in Avalonia.Samples as it does basically similar thing.

**Reply by ColtonHoksbergen** [reply]
It is.  I will grab that and take a look at it now, thank you.

**Reply by ColtonHoksbergen** [reply]
My apologies, I think it was because at some point I got a little lost and my MainWindowViewModel inherited ObservableCollection, not realizing that ViewModelBase inherts ObservableCollection for me. At least that's the difference I noticed in my code versus the ToDoList code.

**Reply by timunie** [reply]
Don't mess up ObservableObject - which is the base class and ObservableCollection - which is for items 

https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.observablecollection-1?view=net-8.0
